### PR TITLE
interfaces: add acrn interface for launching virtual machines on ACRN

### PR DESCRIPTION
**Pull request by ttzeng**
This commit adds a new interface for launching VMs on ACRN.

[ACRN](https://projectacrn.org/) is a Type 1 hypervisor designed for embedded devices, which provides a lightweight, real-time, and safety-critical virtualization environment for launching virtual machines on bare-metal devices. We have packaged the [hypervisor](https://github.com/projectacrn/acrn-hypervisor), [ACRN kernel](https://github.com/projectacrn/acrn-kernel), and utilities in the ACRN gadget and kernel snaps in the [acrn-snap](https://github.com/gvancuts/acrn-snap) repository for launching VMs on Ubuntu Core, but still requires some apparmor permissions to access the required resources.

Signed-off-by: Tonny Tzeng <tonny.tzeng@intel.com>

### CODE REVIEW

**Comment by ttzeng** [comment]
@gvancuts @alfonsosanchezbeato while creating this PR to propose changes required for running VMs on ACRN, it seems at least a reviewer must be specified manually. Could you advise how to kick off the review process? Thanks.

File: interfaces/builtin/acrn.go
Comment (bboozzoo) [inline interfaces/builtin/acrn.go:35]: This will probably need an udev rule.

Comment (bboozzoo) [inline interfaces/builtin/acrn.go:35]: Are there any special modules needed for this device? In which case, they may be a need to override KModConnectedPlug() or set connectedPlugKModModules

Comment (ttzeng) [reply interfaces/builtin/acrn.go:35]: @bboozzoo thanks for the review comments.
As the VHM module is static compiled in ACRN kernel, do we still require udev rules?

Comment (ttzeng) [reply interfaces/builtin/acrn.go:35]: @bboozzoo is there wiki pages documenting these callbacks? I learn it from the `kvm.go`, it looks to me the [KModConnectedPlug()](https://github.com/snapcore/snapd/blob/master/interfaces/builtin/kvm.go#L78-L101) is used to check the kernel module presence depending on the CPU type. But in our case, the `/dev/acrn_vhm` is registered by a [static linked](https://github.com/projectacrn/acrn-kernel/blob/master/kernel_config_uefi_sos#L771) kernel driver, do we need to provide an empty callback? Thanks.

Comment (bboozzoo) [reply interfaces/builtin/acrn.go:35]: Try running `udevadm info /dev/acrn_vhm`. Is there any information about the device? The udev rules are used to tag devices which can be accessed by given snap. When the snap application runs we set up a device cgroup and allow access to select devices only.

Comment (bboozzoo) [reply interfaces/builtin/acrn.go:35]: Is it possible to build it as a dynamically loaded module? Does it have aliases set up such that it will get automatically loaded?

Comment (ttzeng) [reply interfaces/builtin/acrn.go:35]: @bboozzoo thanks for your quick response. Below is the output from `udevadm info /dev/acrn_vhm`:
```
P: /devices/virtual/vhm/acrn_vhm
N: acrn_vhm
E: DEVNAME=/dev/acrn_vhm
E: DEVPATH=/devices/virtual/vhm/acrn_vhm
E: MAJOR=243
E: MINOR=0
E: SUBSYSTEM=vhm
E: USEC_INITIALIZED=3905750
E: net.ifnames=0
```
Would you advise what should we put in `acrnConnectedPlugUDev`?

Comment (ttzeng) [reply interfaces/builtin/acrn.go:35]: @bboozzoo , the ACRN VHM driver does not support kernel module dynamic loading per this [Kconfig](https://github.com/projectacrn/acrn-kernel/blob/master/drivers/vhm/Kconfig). Any security concern? @gvancuts may provide additional clarification.

File: interfaces/builtin/hugepages_control.go
Comment (anonymouse64) [inline interfaces/builtin/hugepages_control.go:46]: can you show what sort of denials you see with this? we try very hard to restrict mounts as much as possible with options or locations, since they are usually rife with sandbox escape mechanisms

Comment (anonymouse64) [reply interfaces/builtin/hugepages_control.go:46]: err rather what sort of denials you see without this?

Comment (ttzeng) [reply interfaces/builtin/hugepages_control.go:46]: @anonymouse64 , thanks for review. Below is the message we got from the `snappy-debug` without this mount permission:
```
= AppArmor =
Time: May 22 01:27:14
Log: apparmor="DENIED" operation="mount" info="failed srcname match" error=-13 profile="snap.acrn.launch-uos" name="/var/snap/acrn/common/run/hugepage/huge_lv1/vm-clear-34500-kvm/" pid=1548 comm="acrn-dm" fstype="hugetlbfs" srcname="none"
```
From the ACRN source code, the program failed at [init_hugetlb()](https://github.com/projectacrn/acrn-hypervisor/blob/master/devicemodel/core/main.c#L1015), and bisect the issue indicates it was caused by this [mount](https://github.com/projectacrn/acrn-hypervisor/blob/master/devicemodel/core/hugetlb.c#L443) operation. So we'd like to add hugetlbfs mount permission to the `hugepages-control` interface, because it should be hugepages MMIO related. Thanks.

File: interfaces/builtin/acrn.go
Comment (gvancuts) [reply interfaces/builtin/acrn.go:35]: /cc @shuox to help answer questions about the ACRN kernel driver. The version we use can only be statically compiled in the kernel. The version that was accepted upstreamed more recently (see [this blog](https://projectacrn.org/acrn-hypervisor-service-module-upstreamed-to-linux-kernel/) for additional details) can be configured as a module, but we are are not using this yet with the user-space components from https://github.com/projectacrn/acrn-hypervisor.

Comment (ttzeng) [reply interfaces/builtin/acrn.go:35]: @bboozzoo , I've referenced the `time-control` interface and add an udev rule for it, would you please advice whether it's applicable? Thanks.

Comment (shuox) [reply interfaces/builtin/acrn.go:35]: Yes. As @gvancuts mentioned, VHM driver will be drop (and it had never be upstreamed). The new driver named HSM (derived from VHM) has been accepted by Linux 5.12

The HSM use '/dev/acrn_hsm' as node. And it can be compiled as a module.

Comment (anonymouse64) [reply interfaces/builtin/acrn.go:35]: Is there any reason to not also include `/dev/acrn_hsm` in the AppArmor snippet for this interface and mention the module name in the connectedPlugKModModules as well? AIUI, for your deployment you will have a statically linked kernel module, but going forward in the future others may wish to use this and may be using the dynamically loaded module in 5.12.

File: interfaces/builtin/hugepages_control.go
Comment (anonymouse64) [reply interfaces/builtin/hugepages_control.go:46]: We should further restrict this rule to specify the srcname as `none` and the destination (`name`) to something like $SNAP_DATA or $SNAP_COMMON 

```suggestion
mount fstype=hugetlbfs none -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/** ,
```

File: interfaces/builtin/acrn.go
Comment (anonymouse64) [inline interfaces/builtin/acrn.go:38]: this is already in network-control, I don't think we should include this here, if your application/snap needs to configure network devices it should use the `network-control` interface alongside this `acrn` interface.

File: interfaces/builtin/acrn.go
Comment (ttzeng) [reply interfaces/builtin/acrn.go:40]: @anonymouse64 . Thank you for feedbacks. I will include the new device node in the apparmor snippet and add the `connectedPlugKModModules()` in the amended commit.

File: interfaces/builtin/hugepages_control.go
Comment (ttzeng) [reply interfaces/builtin/hugepages_control.go]: @anonymouse64 , thanks for coach, I've incorporated the suggested change to the commit.

File: interfaces/builtin/acrn.go
Comment (ttzeng) [reply interfaces/builtin/acrn.go:44]: @anonymouse64 , I did try to connect the `network-control` interface, but it looks like there may some configurations blocking our accesses to the `/dev/acrn_vhm`, so I add the required `/dev/net/tun` permission to the snippet directly.

Comment (anonymouse64) [reply interfaces/builtin/acrn.go:44]: Did you use snapd built from this branch to get the acrn and the network-control interfaces at the same time? perhaps @alfonsosanchezbeato can help you debug this? There shouldn't be anything in network-control that conflicts with this new acrn interface providing access to `acrn_vhm`.

**Comment by ttzeng** [comment]
I got a PR run failure notification which indicates the `unit-tests` jobs failed with the following error: 
```
Missing high-level test for interface 'acrn'. Please add to:
* tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
* tests/lib/snaps/test-snapd-policy-app-provider-core/meta/snap.yaml (if needed)
* tests/lib/snaps/test-snapd-policy-app-provider-classic/meta/snap.yaml (if needed)
```
It looks like `acrn` should be added to the above yaml file. Am I responsible for patching the yaml file? or what process should I go through to add it in? Thanks.

**Comment by anonymouse64** [comment]
@ttzeng yes please if you are able to figure out the changes to make that would be great, but if you can't get it working then we can push the required changes here.

Also, fwiw have you tried running the tests locally with `./run-checks` ? I would recommend doing the following:

```
SKIP_GOFMT=1 ./run-checks --static
./run-checks --unit
```

**Comment by ttzeng** [comment]
> Also, fwiw have you tried running the tests locally with `./run-checks` ? I would recommend doing the following:
> 
> ```
> SKIP_GOFMT=1 ./run-checks --static
> ./run-checks --unit
> ```
Thanks for coaching, I can replicate the error with the above command locally, so I've patched the yaml file to pass the unit tests. Please help to review and restart the workflow. Thanks!

File: interfaces/builtin/acrn.go
Comment (anonymouse64) [inline interfaces/builtin/acrn.go:2]: this also should be Copyright Canonical

**Review by anonymouse64 — CHANGES_REQUESTED** [review]
policy looks fine now, but a couple of things need to be addressed:

- [ ] the license for the interface is wrong, it needs to be GPL
- [ ] the Copyright should be Canonical
- [ ] The interface needs to be super-privileged, see my comments for what needs to change for this
- [ ] after changing to super-privileged you will likely need to adjust some more tests, let me know if it's unclear which bits need to be changed, but the test failure should make it clear what needs to be changed
- [ ] The name of the interface needs to be approved by @pedronis 
- [ ] The Canonical Security team needs to approve the policy here, this will be done soon since we have tagged the pr "needs security review"

**Comment by ttzeng** [comment]
> * [ ]  the license for the interface is wrong, it needs to be GPL
> * [ ]  the Copyright should be Canonical
> * [ ]  The interface needs to be super-privileged, see my comments for what needs to change for this
> * [ ]  after changing to super-privileged you will likely need to adjust some more tests, let me know if it's unclear which bits need to be changed, but the test failure should make it clear what needs to be changed
> * [ ]  The name of the interface needs to be approved by @pedronis
> * [ ]  The Canonical Security team needs to approve the policy here, this will be done soon since we have tagged the pr "needs security review"

@anonymouse64 , thank you very much for providing review feedbacks, I'll address them as soon as I can. While going through the list, I'd like to clarify on the copyright: it make sense to GPLed the interface as it will be static linked in the `snapd` binary, but I am not sure whether we could put the code beyond company license, it will need to consult with legal?

**Comment by anonymouse64** [comment]
>  I am not sure whether we could put the code beyond company license, it will need to consult with legal?

I would suggest you reach out to your legal department at Intel, but actually I think I was wrong about needing copyright to be Canonical, in any case please have your legal department look into https://ubuntu.com/legal/contributors/faq before making a final decision. I think section `Section 2.1(a)` of the agreement actually states that you own the copyright but that you grant Canonical a license to use it.

**Comment by anonymouse64** [comment]
Marking blocked until the license and copyright situation is sorted out

**Comment by pedronis** [comment]
superseded by #11097 (https://github.com/snapcore/snapd/pull/11097#issuecomment-993524442)

**Pull request closed without merging**
